api/envoy/config/core/v3/substitution_format_string.proto

```diff
@@ -61,6 +64,12 @@ message SubstitutionFormatString {
     //  }
     //
     google.protobuf.Struct json_format = 2 [(validate.rules).message = {required: true}];
+
+    // Specify a format with command operators, similar to text_format, but from a DataSource.
+    DataSource text_format_source = 5;
```

Comment:
It seems that this should not be the place to add DataSource
1) why it only apply to text_format, why not json_format
2) data_source has `inline` which is same as `text_format` here.

Is it possible to specify in the file_access_log config to DataSource the whole SubstitutionFormatString?

Reply:
This only applies to text_formart because json_format is, as I understand it, a convenience config for writing structured config inline to represent json text. For example:

```
json_format:
  x: 1
  y: 2
```

To supply json as text_format_source, one would write JSON text into the source and set content_type appropriately. Does this match your understanding? If we were to support a DataSource for json_format, what would be your preferred way of using it?

Reply:
The question still exsit, what the rationale behind adding this as data_source, usually DataSource is used when the content is expected to be large, or it cannot fit in a protobuf string (i.e. binary). I don't see either of them apply?

Reply:
Agreed on the usage of DataSource. The rationale is that the text may be relatively large (e.g. a static html page) and may not be suitable for inline strings. The use case that comes to mind is a local reply or response map configuration that serves a variety of different html pages for different response codes.

Reply:
I see that makes sense.

If that's the case, I think we want to deprecate `text_format` and use `text_format_source` as alternative, since DataSource has inline_string, then we should leave `fixed_text`.

Reply:
> Is it possible to specify in the file_access_log config to DataSource the whole SubstitutionFormatString?

I don't see a reason it would be impossible. but for log format that's relatively small, there's no point to have it in a file.

Reply:
If we go with this way,  someone wants to read json_format from filter,  we would need to add json_format_source.  This will be bad.

Reply:
Someone can still use `json_format` if they want to specify an inline representation of their JSON for filter config. If they want to specify a large JSON string from a DataSource, they'd be able to use `text_format_source` with the JSON in the file and then set `content_type` to `application/json`. I don't think `json_format_source` is necessary. What do you think?

Reply:
> I see that makes sense.
> 
> If that's the case, I think we want to deprecate `text_format` and use `text_format_source` as alternative, since DataSource has inline_string, then we should leave `fixed_text`.

I agree that if we go this route, then long term we do not need text_format since the DataSource variant supports inline strings via `inline_string`. If we did not relax the validation in `inline_string` to accept an empty string (it does not currently) then we'd need to indeed add `fixed_text` to support that case.

Reply:
I see.  I am fine with replacing `text_format` with `text_format_source`.
For `fixed_text`, it seems that its only use case is empty string since data_source.inline_string doesn't allow it.   Maybe change this field to `bool  empty`?

Reply:
I'm ok with a field for `empty`, though I think have a slight preference for `empty_format` which represents an empty format, in the same way that `text_format` represents the (non-empty) text format. @lizan originally commented on that here https://github.com/envoyproxy/envoy/pull/14221#discussion_r536029973 and suggested relaxing the constraints on an existing field, though I'm more of a fan of adding a new explicit field and preserving all existing behaviors. There could be some control plain implementations that depend on the existing field validation. I don't know how strong of an argument that is, though.

I'll vote for `bool empty_format`. @lizan / @qiwzhang final thoughts/votes?

Reply:
Taking a step back, if you want empty, why would you specify a [`body_format`](https://github.com/envoyproxy/envoy/pull/14221/files#diff-8d3ae8065707ddf8386a198222c3872ef85eeaf282fabeacbe7a15b37da3d501R88), the field itself can be null?

Reply:
I feel like a null SubstitutionFormatString could be interpreted as "no substitution will take place" as opposed to "substitute with nothing". What do you think?

Reply:
> I feel like a null SubstitutionFormatString could be interpreted as "no substitution will take place" as opposed to "substitute with nothing". What do you think?

It depends on the context where this is used, but that seems natural.

So, after thinking a bit, my vote will be relaxing the constraint on what we have for `text_format` and/or `inline_string`/`inline_bytes`, because you can also have an empty file be specified as DataSource to make it empty, so that way it is more consistent.

Reply:
I sided with @lizan .  Adding extra fields may cause confusion, e.g. fixed text has to be specified in `text_format_source.inline_string`, but an empty one has to be specified in `empty_format`.

Reply:
Sounds good. I will relax the constraint on `text_format` and remove `fixed_text`.

Reply:
Let's also deprecate `text_format` in favor of `text_format_source`.

Reply:
Can do. For deprecations, is it appropriate to mark a field deprecated in `v3` and remove it in `v4alpha` or is there a different process?

Reply:
Yes, just mark `text_format` field deprecated and run the proto_format script, it will take care of the rest.

Reply:
Thanks. I should relax the field validation for `inline_string` on `DataSource` too, right?

Reply:
That will have larger impact and I think we should do it in a separate PR, at least without it we can unblock #14221.

Reply:
Great. Thanks!